### PR TITLE
feat(doorstop): Implement external parent

### PR DESCRIPTION
**Pull request by lbiaggi**
We added an implementation that pull a remote parent for a doorstop document if not present.

It also validates if you try to import a prefix twice with different tags.

- This Implementation should be capable of doing recursive pulling as well (not tested yet)

### CODE REVIEW

**Comment by devcurmudgeon** [comment]
A few comments:

- I think this should be "external-parent", not necessarily remote
- Ideally path/url could also be a local path, even a relative path to an existing directory outside the current repo (e.g. for local testing)
- It's not clear how to use this on a remote repo that requires authentication
- I see "WARNING: no working copy found at:..." even after this has created a working copy
- I see quite a few blank spaces at the ends of lines in the patch
- when running the `doorstop` command (no arguments) with valid external parent set, I don't see any printout of the document tree (whereas the tree is printed with no external parent)

**Comment by frankmillward** [comment]
Adding an external parent to a (child) doorstop project invalidates all references to files in the child.

i.e. references in the child of form:
```yaml
references:
 - path: path/to/file.txt
   type: file
```

**Comment by lbiaggi** [comment]
> A few comments:
> 
>     * I think this should be "external-parent", not necessarily remote
I agree, but Doorstop relies heavily on git tree and strings to deal with paths, which IMO is not ideal. I would like first to move it to use  `pathlib` before implementing something like this. The remote parent already broke “external references”.

> 
>     * Ideally path/url could also be a local path, even a relative path to an existing directory outside the current repo (e.g. for local testing)
> 
>     * It's not clear how to use this on a remote repo that requires authentication
In my vision, would be related to the permission from CI/ USER.  I'm using against authenticated repos like this without any issue.
> 
>     * I see "WARNING: no working copy found at:..." even after this has created a working copy
I will check, as I'm not seeing this
> 
>     * I see quite a few blank spaces at the ends of lines in the patch
Can you detail more about it?
> 
>     * when running the `doorstop` command (no arguments) with valid external parent set, I don't see any printout of the document tree (whereas the tree is printed with no external parent)

Probably something is wrong about the setup? Because it shouldn't be possible to print a tree without the parent. Can you provide the steps to reproduce?

**Comment by lbiaggi** [comment]
> Adding an external parent to a (child) doorstop project invalidates all references to files in the child.
> 
> i.e. references in the child of form:
> 
> ```yaml
> references:
>  - path: path/to/file.txt
>    type: file
> ```

Argh, probably it is because those paths are just strings and doorstop it is relying on the git tree to find those files and importing a project in another folder is breaking its logic.

**Comment by KasparMatas** [comment]
The commit title should be feat(... not eat(...

**Comment by frankmillward** [comment]
Found the following bug locally, would be interested if you can reproduce this @lbiaggi:

Once a git tag is added to the definition of the remote parent
```
remote_parent_url: git@myprojects.com:path/to/project
remote_parent_tag: main
```
and `doorstop` is run, changing the tag in the yaml tag definition has no effect, but no error is reported, even if that tag does not exist in the remote.

As a workaround, the `.doorstop_remote` folder can be manually deleted, but this ideally would be fixed or at the least well documented.

**Comment by lbiaggi** [comment]
> Found the following bug locally, would be interested if you can reproduce this @lbiaggi:
> 
> Once a git tag is added to the definition of the remote parent
> 
> ```
> remote_parent_url: git@myprojects.com:path/to/project
> remote_parent_tag: main
> ```
> 
> and `doorstop` is run, changing the tag in the yaml tag definition has no effect, but no error is reported, even if that tag does not exist in the remote.
> 
> As a workaround, the `.doorstop_remote` folder can be manually deleted, but this ideally would be fixed or at the least well documented.

Thanks @frankmillward, I will add a --clean-remote option to purge and document its usage for now.

File: docs/reference/document.md [hunk 1]
```diff
@@ -103,4 +103,28 @@ In `path/to/file.yml`:
...
   have several
   lines.
-```
\ No newline at end of file
+```
+# Beta Features
+
+## External documents
+
+**Note: When you use this feature, it is designed assuming that the external parent will be responsible to identify the root document.**
```
Comment (reiterative) [inline docs/reference/document.md]:
```suggestion
**Note: When you use this feature, it is designed assuming that the external parent will be responsible for identifying the root document.**
```

File: docs/reference/document.md [hunk 2]
```diff
@@ -103,4 +103,28 @@ In `path/to/file.yml`:
...
+external_parent: <GIT URL> <-- git url to used to pull the external parent **Required**
+external_parent_tag: <GIT_TAG> <-- tag or branch to be used **REQUIRED**
+settings:
+  parent: REQ <---- PREFIX from the external document **REQUIRED**
+  digits: 3
+  prefix: REQ
+  sep: ''
+```
+
+If for some reason you needed to update external tag or parent location, run doorstop with the flag `--clean-external-docs`.
```
Comment (reiterative) [inline docs/reference/document.md]:
```suggestion
If you need to update the external tag or parent location, run doorstop with the flag `--clean-external-docs`.
```

File: doorstop/core/vcs/base.py [hunk 3]
```diff
@@ -115,8 +115,11 @@ def paths(self):
                         log.debug(f"not skipping: {relpath}")
                         self._path_cache.append((path, filename, relpath))
                         continue
-                    # Skip hidden paths
-                    if os.path.sep + "." in os.path.sep + relpath:
+                    # Skip hidden paths except for 'remote folder'
```
Comment (reiterative) [inline doorstop/core/vcs/base.py]: what if we want to include hidden files in references? For example, if we reference a CI file in a `.gitlab-ci` folder.

Comment (lbiaggi) [reply doorstop/core/vcs/base.py]: you would still have the  `.unhidden` bit.

Comment (lbiaggi) [reply doorstop/core/vcs/base.py]: https://github.com/CodethinkLabs/doorstop/pull/15/files#diff-5fa9aba911bf89d1fddf65b8e71f3a6b5340aeb9f14e056439a988060dc46983R114

Comment (reiterative) [reply doorstop/core/vcs/base.py]: Doh! I missed that, because it wasn't shown in the 'changed' snippet.

**Review by daplm — APPROVED** [review]
I did not test anything, But I read through the changes and did not find any obvious errors.

**Review by devcurmudgeon — APPROVED** [review]

**Pull request merged**